src/ripple/app/ledger/LedgerEntrySet.cpp

```diff
@@ -1024,52 +1024,66 @@ uint256 LedgerEntrySet::getNextLedgerIndex (
 
 void LedgerEntrySet::incrementOwnerCount (SLE::ref sleAccount)
 {
+    increaseOwnerCount (sleAccount, 1);
+}
+
+void LedgerEntrySet::incrementOwnerCount (Account const& owner)
+{
+    increaseOwnerCount(
+        entryCache (ltACCOUNT_ROOT, getAccountRootIndex (owner)), 1);
+}
+
+void
+LedgerEntrySet::increaseOwnerCount (SLE::ref sleAccount, std::size_t howMuch)
+{
     assert (sleAccount);
 
     std::uint32_t const current_count = sleAccount->getFieldU32 (sfOwnerCount);
+    std::uint32_t new_count = current_count + howMuch;
 
-    if (current_count == std::numeric_limits<std::uint32_t>::max ())
+    // Check for integer overflow -- well defined behavior on unsigned.
+    if (new_count < current_count)
     {
         WriteLog (lsFATAL, LedgerEntrySet) <<
             "Account " << sleAccount->getFieldAccount160 (sfAccount) <<
             " owner count exceeds max!";
-        return;
+        new_count = std::numeric_limits<std::uint32_t>::max ();
     }
-
-    sleAccount->setFieldU32 (sfOwnerCount, current_count + 1);
+    sleAccount->setFieldU32 (sfOwnerCount, new_count);
     entryModify (sleAccount);
 }
 
-void LedgerEntrySet::incrementOwnerCount (Account const& owner)
+void LedgerEntrySet::decrementOwnerCount (SLE::ref sleAccount)
 {
-    incrementOwnerCount(entryCache (ltACCOUNT_ROOT,
-        getAccountRootIndex (owner)));
+    decreaseOwnerCount (sleAccount, 1);
 }
 
-void LedgerEntrySet::decrementOwnerCount (SLE::ref sleAccount)
+void LedgerEntrySet::decrementOwnerCount (Account const& owner)
+{
+    decreaseOwnerCount(
+        entryCache (ltACCOUNT_ROOT, getAccountRootIndex (owner)), 1);
+}
+
+void
+LedgerEntrySet::decreaseOwnerCount (SLE::ref sleAccount, std::size_t howMuch)
 {
     assert (sleAccount);
 
     std::uint32_t const current_count = sleAccount->getFieldU32 (sfOwnerCount);
+    std::uint32_t new_count = current_count - howMuch;
 
-    if (current_count == 0)
+    // Check for integer underflow -- well defined behavior on unsigned.
+    if (new_count > current_count)
     {
         WriteLog (lsFATAL, LedgerEntrySet) <<
             "Account " << sleAccount->getFieldAccount160 (sfAccount) <<
-            " owner count is already 0!";
-        return;
+            " owner count set below 0!";
+        new_count = 0;
```

Comment:
I can say with 95% confidence the existing npm tests would have triggered the assert and the bug would have been found sooner.  The tests DID exercise this code, though failed to exhaustively assert expected behaviour for all the changes.